programs/evm_loader/src/processor.rs

```diff
@@ -0,0 +1,231 @@
+                        }
+                    }
+
+                    if !processed {
+                        return Err(InstructionError::NotEnoughAccountKeys)
+                    }
+                },
+            }
+        }
+
+        Ok(())
+    }
+
+    fn pubkey_to_address(key: &Pubkey) -> H160 {
+        H256::from_slice(key.as_ref()).into()
+    }
+
+    pub fn do_write(
+        keyed_accounts: &mut [KeyedAccount],
+        offset: u32,
+        bytes: &[u8],
+    ) -> Result<(), InstructionError> {
+        let mut keyed_accounts_iter = keyed_accounts.iter_mut();
+        let keyed_account = next_keyed_account(&mut keyed_accounts_iter)?;
+
+        if keyed_account.signer_key().is_none() {
+            debug!("Error: key[0] did not sign the transaction");
+            return Err(InstructionError::MissingRequiredSignature);
+        }
+        let offset = offset as usize;
+        let len = bytes.len();
+        trace!("Write: offset={} length={}", offset, len);
+        if keyed_account.account.data.len() < offset + len {
+            debug!(
+                "Error: Write overflow: {} < {}",
+                keyed_account.account.data.len(),
+                offset + len
+            );
+            return Err(InstructionError::AccountDataTooSmall);
+        }
+        keyed_account.account.data[offset..offset + len].copy_from_slice(&bytes);
+        Ok(())
+    }
+
+    pub fn do_finalize(keyed_accounts: &mut [KeyedAccount]) -> Result<(), InstructionError> {
+        let mut keyed_accounts_iter = keyed_accounts.iter_mut();
+        let keyed_account = next_keyed_account(&mut keyed_accounts_iter)?;
+        let keyed_address = keyed_account.signer_key().map(|k| Self::pubkey_to_address(k))
+            .ok_or(InstructionError::MissingRequiredSignature)?;
+
+        let vicinity = MemoryVicinity {
+            gas_price: U256::zero(),
+            origin: H160::default(),
+            chain_id: U256::zero(),
+            block_hashes: Vec::new(),
+            block_number: U256::zero(),
+            block_coinbase: H160::default(),
+            block_timestamp: U256::zero(),
+            block_difficulty: U256::zero(),
+            block_gas_limit: U256::zero(),
+        };
+        let code = keyed_account.account.data.clone();
+        let mut state = BTreeMap::new();
+
+        for state_account in keyed_accounts_iter {
+            let (address, account) = Self::read_account(&state_account.account.data)?;
+
+            state.insert(address, account);
+        }
+        let backend = MemoryBackend::new(&vicinity, state);
+        let config = evm::Config::istanbul();
+        let mut executor = StackExecutor::new(&backend, usize::max_value(), &config);
+        let exit_reason =
+            executor.transact_create(keyed_address, U256::zero(), code, usize::max_value());
+
+        let (applies, _logs) = executor.deconstruct();
+
+        if exit_reason.is_succeed() {
+            keyed_account.account.executable = true;
```

Comment:
If there is any other verification of the program that we can do it should be done before its marked `executable`.  The goal is for `do_invoke_main` to only fail due to inputs to the finalized program and not due to the finalized program itself